# Quantitative-Biology-Portfolio

The projects are part of the course Introduction to Quantitative Biology, offered to 4th semester undergraduates at IIITD.

- ### [Protein Sequence Alignment Using Needlemanâ€“Wunsch algorithm](https://github.com/adityasaini70/Quantitative-Biology-Portfolio/tree/master/Protein%20Sequence%20alignment%20using%20NW%20algorithm) 
  The code takes input 2 sequences and aligns them with each other, such that their similarity score is maximum using the **Needleman-Wunsch algorithm**. 

  **Note :** The implementation corresponds to **gap penalty = 0** and the scoring scheme as identity scoring scheme.


  To use the script:

  ```
  python alignment_script.py -i <input_file>
  ```

- ### [Prediction of ATP interacting residues](https://github.com/adityasaini70/Quantitative-Biology-Portfolio/tree/master/Prediction%20of%20ATP%20interacting%20residues)
  The objective of this project was to detect the presence of ATP interacting residues in a given protein sequence.
  - Research Paper referred - **Jagat S Chauhan, Nitish K Mishra & Gajendra PS Raghava : 
    [Identification of ATP binding residues of a protein from its primary sequence](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-10-434)**
