# Quantitative-Biology

The projects are part of the course Introduction to Quantitative Biology, offered to 4th semester undergraduates at IIITD.

- ### [Protein Sequence Alignment Using Needlemanâ€“Wunsch algorithm](https://github.com/adityasaini70/Quantitative-Biology-Portfolio/tree/master/Protein%20Sequence%20alignment%20using%20NW%20algorithm) 
  The code takes input 2 sequences and aligns them with each other, such that their similarity score is maximum using the **Needleman-Wunsch algorithm**. 

  **Note :** The implementation corresponds to **gap penalty = 0** coupled with an identity scoring scheme.


  To use the script:

  ```
  python alignment_script.py -i <input_file>
  ```
  
  **Example Output:**
  
  ![image](https://user-images.githubusercontent.com/49980787/109796994-de0ad300-7c3e-11eb-9e11-af2c345e1a02.png)

- ### [Prediction of ATP interacting residues](https://github.com/adityasaini70/Quantitative-Biology-Portfolio/tree/master/Prediction%20of%20ATP%20interacting%20residues)
  The objective of this project was to detect the presence of ATP interacting residues in a given protein sequence.
  - **Research Paper referred** - Chauhan, J.S., Mishra, N.K. & Raghava, G.P. Identification of ATP binding residues of a protein from its primary sequence. BMC Bioinformatics 10, 434 (2009). https://doi.org/10.1186/1471-2105-10-434
